### PR TITLE
invmod(n::BitInteger): efficient native modular inverses

### PR DESCRIPTION
**Pull request by StefanKarpinski**
Implement algorithm described in https://arxiv.org/pdf/2204.04342.pdf. The algorithm is pleasingly simple and efficient and the generic Julia implementation is also really enjoyable.

### CODE REVIEW

File: base/intfuncs.jl [hunk 1]
```diff
@@ -257,6 +257,33 @@ function invmod(n::Integer, m::Integer)
...
+convenience of not needing to specify a modulus, it is especially inconvenient
+to specify a modulus of `2^N` since this value can never be represented in `T`.
+The modular inverse can also be computed much more efficiently than the general
+case using the algorithm described in https://arxiv.org/pdf/2204.04342.pdf.
+
+!!! compat "Julia 1.11" The single-argument method of `invmod` requires Julia
+    1.11 or later.
+"""
+function invmod(n::T) where {T<:BitInteger}
+    isodd(n) || throw(DomainError(n, "Argument mut be odd."))
```
Comment (JeffBezanson) [inline base/intfuncs.jl:277]:
```suggestion
    isodd(n) || throw(DomainError(n, "Argument must be odd."))
```

**Comment by oscardssmith** [comment]
I'm not sure this is `invmod`. I think this method should possibly be `invmod(n::T, T) where T` for consistency

**Comment by StefanKarpinski** [comment]
It's a modular inverse where the modulus is implied by the type. I could _maybe_ see having `inv(n::T)` do this, but we've already defined that to do `1/n` which is not this operation. I'm certainly ok with having an `invmod(n::Integer, ::Type{T}) where {T<:Integer}` method, but its definition would simply be `invmod(n % T)`. I can add that method to the PR.

**Comment by oscardssmith** [comment]
my point is that I don't think `invmod` makes sense as a 1 argument function (because `inv` mod what?)

**Comment by StefanKarpinski** [comment]
> because inv mod what?

Mod the modulus implied by there type of the argument. It seems reasonable to have an abbreviation for `invmod(n, typeof(n))`; what other function would you add that method to? We could define a new `modinv(n)` function but why? This seems like a very natural and convenient place for this functionality.

**Comment by oscardssmith** [comment]
fair enough.

**Comment by StefanKarpinski** [comment]
Failures look unrelated, this is a simple PR, and there's a decent spread of successes, so I'm gonna merge.

**Pull request merged**